File: send-invoice-without-days-until-due/python/stripe_days_until_due.py
"""Report Stripe send_invoice subscriptions that write invoices with no due date.

Read only. Two paginated GETs and no writes: give this a RESTRICTED key with read
access to Subscriptions and Invoices. The repair is printed, never performed,
because changing payment terms changes what a customer owes and when.
"""
import argparse
import logging
import os
import sys

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("stripe_days_until_due")

API = "https://api.stripe.com/v1"

# Roughly how many days each recurring interval is worth. Used only to compare
# payment terms against the billing period, so month lengths do not matter.
INTERVAL_DAYS = {"day": 1, "week": 7, "month": 30, "year": 365}


def verdict(collection_method, days_until_due, interval_days, undated_open_invoices):
    """Classify one subscription. Pure, so the rules can be tested without a network.

    `days_until_due` is the raw field: None when absent, and 0 is a real value
    meaning due on receipt. `interval_days` is the billing period in days, or
    None if it could not be read. `undated_open_invoices` is how many open
    invoices this subscription already has with a null due_date.

    Returns (state, detail).
    """
    if collection_method != "send_invoice":
        return ("automatic",
                "collection_method is %r: Stripe charges the payment method on "
                "file, so days_until_due does not apply" % (collection_method,))
    if days_until_due is None:
        if undated_open_invoices:
            return ("undated",
                    "days_until_due is null and %d open invoice(s) already have "
                    "due_date null: nothing can mark them overdue"
                    % undated_open_invoices)
        return ("unanchored",
                "days_until_due is null, so every invoice this subscription "
                "writes will have due_date null and can never age")
    if days_until_due == 0:
        return ("on-receipt",
                "net 0, due on receipt: a real term, not a missing one")
    if interval_days and days_until_due >= interval_days:
        return ("overlapping",
                "net %d on a %d day billing period: the next invoice is issued "
                "before this one is due" % (days_until_due, interval_days))
    return ("dated",
            "net %d; due_date is set and the past due machinery has something "
            "to measure from" % days_until_due)


def get(session, path, **params):
    r = session.get(API + path, params=params, timeout=30)
    if r.status_code == 401:
        raise SystemExit("401 from Stripe: the key is wrong, or is for the other mode")
    r.raise_for_status()
    return r.json()


def page_all(session, path, limit, **params):
    """Page a list endpoint until it runs out or the cap is reached."""
    out = []
    params = dict(params, limit=100)
    while True:
        page = get(session, path, **params)
        data = page.get("data", [])
        out.extend(data)
        if not data or not page.get("has_more") or len(out) >= limit:
            break
        params["starting_after"] = data[-1]["id"]
    return out


def interval_days(sub):
    """Billing period length in days, or None when the price cannot be read."""
    items = (sub.get("items") or {}).get("data") or []
    if not items:
        return None
    recurring = (items[0].get("price") or {}).get("recurring") or {}
    unit = INTERVAL_DAYS.get(recurring.get("interval"))
    if not unit:
        return None
    return unit * (recurring.get("interval_count") or 1)


def undated_open_invoices(session, limit):
    """Count open send_invoice invoices with a null due_date, per subscription.

    The list endpoint has no server-side filter for due_date, so the comparison
    is client side over invoices Stripe has already narrowed by status and
    collection method.
    """
    counts = {}
    for inv in page_all(session, "/invoices", limit,
                        status="open", collection_method="send_invoice"):
        if inv.get("due_date") is None:
            sub = inv.get("subscription")
            if isinstance(sub, dict):
                sub = sub.get("id")
            counts[sub] = counts.get(sub, 0) + 1
    return counts


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--top", type=int, default=20,
                    help="how many individual subscriptions to print")
    ap.add_argument("--max-rows", type=int, default=2000,
                    help="stop paginating each list after this many objects")
    args = ap.parse_args()

    key = os.environ.get("STRIPE_API_KEY")
    if not key:
        log.error("set STRIPE_API_KEY (use a restricted, read-only key)")
        return 2

    s = requests.Session()
    s.headers.update({"Authorization": "Bearer " + key})

    undated = undated_open_invoices(s, args.max_rows)
    subs = page_all(s, "/subscriptions", args.max_rows,
                    collection_method="send_invoice", status="all")
    if not subs:
        log.info("no send_invoice subscriptions for this key's mode")
        return 0

    rows = []
    for sub in subs:
        state, detail = verdict(sub.get("collection_method"),
                                sub.get("days_until_due"),
                                interval_days(sub),
                                undated.get(sub.get("id"), 0))
        rows.append((sub.get("id", "<no id>"), state, detail))

    bad = [r for r in rows if r[1] in ("undated", "unanchored", "overlapping")]
    if not bad:
        log.info("%-11s 0 of %d send_invoice subscription(s) without terms",
                 "clear", len(rows))
        return 0

    log.warning("%-11s %d of %d send_invoice subscription(s) need terms",
                "unterm", len(bad), len(rows))
    for sub_id, state, detail in bad[:args.top]:
        log.warning("  %-11s %s  %s", state, sub_id, detail)
        log.warning("      repair: POST %s/subscriptions/%s  days_until_due=30",
                    API, sub_id)
    if len(bad) > args.top:
        log.warning("  ... and %d more", len(bad) - args.top)
    log.warning("  then Dashboard > Settings > Billing > Invoices: enable the "
                "reminder emails and set the past due subscription action")
    log.warning("  invoices already finalized keep their null due_date; those "
                "need a resend or a write off, one at a time")
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: send-invoice-without-days-until-due/node/stripe-days-until-due.mjs
/**
 * Report Stripe send_invoice subscriptions that write invoices with no due date.
 *
 * Read only. Two paginated GETs and no writes: give this a RESTRICTED key with
 * read access to Subscriptions and Invoices. The repair is printed, never performed.
 */
const API = 'https://api.stripe.com/v1';

// Roughly how many days each recurring interval is worth. Used only to compare
// payment terms against the billing period, so month lengths do not matter.
export const INTERVAL_DAYS = { day: 1, week: 7, month: 30, year: 365 };

/**
 * Classify one subscription. Pure, so the rules can be tested without a network.
 * `daysUntilDue` is the raw field: null when absent, and 0 is a real value.
 */
export function verdict(collectionMethod, daysUntilDue, intervalDays, undatedOpenInvoices) {
  if (collectionMethod !== 'send_invoice') {
    return ['automatic',
      `collection_method is ${JSON.stringify(collectionMethod)}: Stripe charges ` +
      'the payment method on file, so days_until_due does not apply'];
  }
  if (daysUntilDue === null || daysUntilDue === undefined) {
    if (undatedOpenInvoices) {
      return ['undated',
        `days_until_due is null and ${undatedOpenInvoices} open invoice(s) already ` +
        'have due_date null: nothing can mark them overdue'];
    }
    return ['unanchored',
      'days_until_due is null, so every invoice this subscription writes will ' +
      'have due_date null and can never age'];
  }
  if (daysUntilDue === 0) {
    return ['on-receipt', 'net 0, due on receipt: a real term, not a missing one'];
  }
  if (intervalDays && daysUntilDue >= intervalDays) {
    return ['overlapping',
      `net ${daysUntilDue} on a ${intervalDays} day billing period: the next ` +
      'invoice is issued before this one is due'];
  }
  return ['dated',
    `net ${daysUntilDue}; due_date is set and the past due machinery has ` +
    'something to measure from'];
}

/** Billing period length in days, or null when the price cannot be read. */
export function intervalDays(sub) {
  const items = sub.items?.data ?? [];
  if (items.length === 0) return null;
  const recurring = items[0].price?.recurring ?? {};
  const unit = INTERVAL_DAYS[recurring.interval];
  if (!unit) return null;
  return unit * (recurring.interval_count ?? 1);
}

async function get(key, path, params = {}) {
  const url = new URL(API + path);
  for (const [k, v] of Object.entries(params)) url.searchParams.set(k, v);
  const res = await fetch(url, { headers: { Authorization: `Bearer ${key}` } });
  if (res.status === 401) {
    throw new Error('401 from Stripe: the key is wrong, or is for the other mode');
  }
  if (!res.ok) throw new Error(`${res.status} from ${url.pathname}`);
  return res.json();
}

async function pageAll(key, path, limit, params = {}) {
  const out = [];
  const q = { ...params, limit: 100 };
  for (;;) {
    const page = await get(key, path, q);
    const data = page.data ?? [];
    out.push(...data);
    if (data.length === 0 || !page.has_more || out.length >= limit) break;
    q.starting_after = data[data.length - 1].id;
  }
  return out;
}

async function main() {
  const key = (process.env.STRIPE_API_KEY || "dummy-stripe-api-key");
  if (!key) {
    console.error('set STRIPE_API_KEY (use a restricted, read-only key)');
    process.exitCode = 2;
    return;
  }

  const maxRows = Number(process.argv[2] ?? 2000);

  const undated = new Map();
  for (const inv of await pageAll(key, '/invoices', maxRows,
    { status: 'open', collection_method: 'send_invoice' })) {
    if (inv.due_date === null || inv.due_date === undefined) {
      const sub = typeof inv.subscription === 'object' && inv.subscription !== null
        ? inv.subscription.id : inv.subscription;
      undated.set(sub, (undated.get(sub) ?? 0) + 1);
    }
  }

  const subs = await pageAll(key, '/subscriptions', maxRows,
    { collection_method: 'send_invoice', status: 'all' });
  if (subs.length === 0) {
    console.log("no send_invoice subscriptions for this key's mode");
    return;
  }

  const rows = subs.map((sub) => ({
    id: sub.id ?? '<no id>',
    v: verdict(sub.collection_method, sub.days_until_due, intervalDays(sub),
      undated.get(sub.id) ?? 0),
  }));

  const bad = rows.filter((r) => ['undated', 'unanchored', 'overlapping'].includes(r.v[0]));
  if (bad.length === 0) {
    console.log(`${'clear'.padEnd(11)} 0 of ${rows.length} send_invoice subscription(s) without terms`);
    return;
  }

  console.warn(`${'unterm'.padEnd(11)} ${bad.length} of ${rows.length} send_invoice subscription(s) need terms`);
  for (const r of bad.slice(0, 20)) {
    console.warn(`  ${r.v[0].padEnd(11)} ${r.id}  ${r.v[1]}`);
    console.warn(`      repair: POST ${API}/subscriptions/${r.id}  days_until_due=30`);
  }
  if (bad.length > 20) console.warn(`  ... and ${bad.length - 20} more`);
  console.warn('  then Dashboard > Settings > Billing > Invoices: enable the ' +
               'reminder emails and set the past due subscription action');
  console.warn('  invoices already finalized keep their null due_date; those ' +
               'need a resend or a write off, one at a time');
  process.exitCode = 1;
}

// Only run when invoked directly. The test file imports this module, and without
// the guard main() would run there too, fail on the missing key, and set a
// non-zero exit code that fails the whole test file even as every test passes.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}
